services/autoscaling/src/simcore_service_autoscaling/utils/auto_scaling_core.py

```diff
@@ -0,0 +1,76 @@
+import logging
+import re
+from typing import Final
+
+from models_library.generated_models.docker_rest_api import Node
+
+from ..core.errors import Ec2InvalidDnsNameError
+from ..core.settings import ApplicationSettings
+from ..models import AssociatedInstance, EC2InstanceData
+from . import utils_docker
+
+_EC2_INTERNAL_DNS_RE: Final[re.Pattern] = re.compile(r"^(?P<host_name>ip-[^.]+).*$")
+_logger = logging.getLogger(__name__)
+
+
+def node_host_name_from_ec2_private_dns(
+    ec2_instance_data: EC2InstanceData,
+) -> str:
+    if match := re.match(_EC2_INTERNAL_DNS_RE, ec2_instance_data.aws_private_dns):
+        return match.group("host_name")
+    raise Ec2InvalidDnsNameError(aws_private_dns_name=ec2_instance_data.aws_private_dns)
+
+
+async def associate_ec2_instances_with_nodes(
+    nodes: list[Node], ec2_instances: list[EC2InstanceData]
+) -> tuple[list[AssociatedInstance], list[EC2InstanceData]]:
+    """returns the associated and non-associated instances"""
+    associated_instances: list[AssociatedInstance] = []
+    non_associated_instances: list[EC2InstanceData] = []
+
+    def _find_node_with_name(node: Node) -> bool:
```

Comment:
reads better as a bool return
```suggestion
    def _found_node_with_name(node: Node) -> bool:
```